Generics

yana-glt opened the pull request.

yana-glt committed:
Add generics and arrays

yana-glt committed:
Clean up ignored files in generics branch

vboltrukanis commented:
Edit this PR, so that destination branch is `new_methods`. This way we will see only changes related to this PR. Thanks in advance

yana-glt committed:
Change exceptions

yana-glt commented:
Hello,
I understand what you mean. I changed the branch and added a new commit
with changes in Exceptions to generics branch.

вт, 13 дек. 2022 г. в 14:56, Yana Galitsyna ***@***.***>:

> Excuse me, do you mean swap branches here?
>>
>
> вт, 13 дек. 2022 г. в 14:28, vboltrukanis ***@***.***>:
>
>> Edit this PR, so that destination branch is new_methods. This way we
>> will see only changes related to this PR. Thanks in advance
>>
>> —
>> Reply to this email directly, view it on GitHub
>> <https://github.com/yana-glt/Banks/pull/3#issuecomment-1348545929>, or
>> unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/A2PKG7Z7KXGFETH75D2KSYTWNB2WJANCNFSM6AAAAAAS4IWKJQ>
>> .
>> You are receiving this because you authored the thread.Message ID:
>> ***@***.***>
>>
>

vboltrukanis commented on src/branches/Branch.java:
```diff
@@ -98,4 +98,8 @@ public void setListOfClients(ArrayList<Client> listOfClients) {
 		return String.format("Branch [id=%d , name=%s]", id, name);
 	}
+
+	public static <T> T getLastElementOfList(ArrayList<T> list) {
```
Thats a strange method here)

yana-glt replied:
OK. I'll get rid of it)

vboltrukanis commented on src/clients/IndividualClient.java:
```diff
@@ -87,6 +91,14 @@ public void setAccount(IndividualClientsAccount account) {
+	public void setListOfCredits(List<IndividualClientsCredit> listOfCredits) {
+		this.listOfCredits = listOfCredits;
+	}
+
```
addCredit() ?
removeCredit() ?

yana-glt replied:
I implemented adding a new credit to the list of client credits in the giveCredit() method and added the closeCredit() method, in which the credit is removed from the list when the debt is paid in full.

vboltrukanis commented on src/products/accounts/Account.java:
```diff
@@ -86,6 +88,22 @@ public boolean getStatus() {
+		return responsibleExecutorOfBloking;
+	}
+
+	public void setResponsibleExecutorOfBloking(Employee responsibleExecutorOfBloking) {
```
this should be set in setBlock() function. We dont want to allow change of such important data easily.

yana-glt replied:
I don't quite understand what you mean. I already have a blockAccount() method in the Account class that is responsible for blocking. And in it I use the setResponsibleExecutorOfBloking() method to set the responsible executor.

vboltrukanis commented on src/products/accounts/Account.java:
```diff
@@ -119,51 +137,95 @@ public String toString() {
+			logger.warn(String.format("The account %s was blocked on %s. Responsible executor: %s",
+					e.getAccountNumber(), e.getAccountBlockDate(), e.getResponsibleExecutorOfBlocking()), e);
+		}
+		System.out.println(sum);
```
Some more info here maybe.

yana-glt replied:
ok

vboltrukanis commented on src/products/accounts/Account.java:
```diff
@@ -119,51 +137,95 @@ public String toString() {
+		return sum;
+	}
+
+	public static boolean transferMoney(Account srcAccount, Account dstAccount, float transferAmount) {
```
i believe we should handle exceptions from small function here, and use throws in deduceBalance and refillBalance.

yana-glt replied:
I tried to implement this. But it still looks cumbersome, because it is necessary to provide that if the first account is not blocked and money is debited from it, and the second account turns out to be blocked, then it is necessary to roll back the operation on the first account.

vboltrukanis commented on src/products/cards/IndividualClientsCard.java:
```diff
@@ -23,8 +29,17 @@ public void setAccount(IndividualClientsAccount account) {
+		return client;
+	}
+
+	public void setClient(IndividualClient client) {
```
Unlikely we ever will want to change the client of a card, so this method kinda conflicts with business logic.

yana-glt replied:
Agree. It makes sense

yana-glt committed:
Edit account balance change methods, implementation of addCredit(), removeCredit()

vboltrukanis commented on src/clients/CorporateClient.java:
```diff
@@ -1,136 +1,147 @@
+    private String name;
+    private String directorsName;
+    private String directorsSurname;
+    private String directorsPhoneNumber;
```
director is an Employee as far as i understand, so here you can just reference related object, instead of hardcoding this data

yana-glt replied:
The director is an employee of his company, but not of the bank. The classes Branch, Department and Employee refer specifically to the organization of processes in the bank. I would not like to confuse these two hierarchies.

vboltrukanis commented on src/products/cards/Card.java:
```diff
@@ -1,85 +1,106 @@
+                nameAndSurname, expirationDate, cvvCode, issueDate, status);
+    }
+
+    public void determineCardExpirationDate() {
```
this can be private and called inside of cinstructor

yana-glt replied:
I fixed it

vboltrukanis commented on src/products/cards/Card.java:
```diff
@@ -1,85 +1,106 @@
+    }
+
+    public void blockOnRequest() {
+        this.setStatus(false);
```
you can add logger here, as it is important operation.

yana-glt replied:
ok

vboltrukanis commented on src/products/cards/Card.java:
```diff
@@ -1,85 +1,106 @@
+    public void blockAfterExpiration() {
+        LocalDate currentDate = LocalDate.now();
+        if (this.expirationDate.isEqual(currentDate) || this.expirationDate.isBefore(currentDate)) {
+            this.blockOnRequest();
```
Thats strange from a perspective of business logic. Maybe we can send a string like blockReason as a param to make it more understandable.

yana-glt replied:
I added this information to the logger

vboltrukanis reviewed (approved):
Minor changes. Plz review them and fix if you believe this is needed before merge.

yana-glt committed:
Change the Card class, methods blockAfterExpiration() and determineCardExpirationDate()

yana-glt committed:
Merge branch 'main' into generics
